### PR TITLE
fix: Missing face boundary condition support

### LINKED ISSUE #3764: CompositionalMultiphaseHybridFVM: Missing face boundary condition support
The CompositionalMultiphaseHybridFVM solver lacks face-based Dirichlet boundary condition support, unlike SinglePhaseHybridFVM which has complete implementation via applyFaceDirichletBC. 

Impact:
Users cannot apply face-based boundary conditions in compositional multiphase hybrid FVM simulations.

### PR DESCRIPTION
**Pull request by OmarDuran**
Implements face-based Dirichlet boundary condition support for CompositionalMultiphaseHybridFVM (parity with SinglePhaseHybridFVM). It fixes #3764.

To ensure the correctness of the implementation, an integrated test is included. The physical setup consists of an inflow composition entering a unit cube at x=0 (inlet pressure) and exiting at x=1 (outlet pressure).

For a mixed mesh containing hexahedra, pyramids, and tetrahedra (hex_pyr_tet_nested_mixed.vtu), the figure below shows the pressure field as a color map using a shrunk-element representation (note the pyramids highlighted with black edges). When the inconsistent FV-TPFA method is applied, discrepancies arise between the pressure solution and the transported gas and water saturations. For the consistent discretization MFD, the employed inner product is quasiTPFA:
comparison_IP_QTPFA

As expected, when a inconsistent inner product (MFD-TPFA) is used, the solutions should coincide:
comparison_IP_TPFA

For completeness, the test also includes a regular mesh to verify that all methods produce consistent results in that case.

### CODE REVIEW

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp [hunk 1]
```diff
@@ -572,7 +573,104 @@ void CompositionalMultiphaseHybridFVM::applyBoundaryConditions( real64 const tim
...
+        // Apply field value to the lhs and rhs
+        FieldSpecificationEqual::SpecifyFieldValue( dofIndex,
+                                                    rankOffset,
+                                                    localMatrix,
+                                                    rhsValue,
+                                                    presFace[kf],
+                                                    presFace[kf] );
+        localRhs[localRow] = rhsValue;
+      } );
+
```
[the remark this answers is not in the document]
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp:844]: The boundary condition logic is relatively clear within the lambda. No changes needed.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp [hunk 2]
```diff
@@ -572,7 +573,104 @@ void CompositionalMultiphaseHybridFVM::applyBoundaryConditions( real64 const tim
...
+                                                  setName, targetGroup.getName(), numTargetFaces ),
+                                        fs );
+      }
+
+      // Using the field specification functions to apply the boundary conditions to the system
+      fs.applyFieldValue< FieldSpecificationEqual,
+                          parallelDevicePolicy<> >( targetSet,
+                                                    time_n + dt,
+                                                    targetGroup,
+                                                    flow::facePressure::key() );
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp]: how about compositions and temperature?

**Review by victorapm — APPROVED** [review]

**Review by paveltomin — APPROVED** [review]

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp (hunk 2 above)
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp]: I have included those boundary conditions and added an integrated test to ensure that the same approximation is obtained with both FV-TPFA and MFD-TPFA discretizations.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp [hunk 3]
```diff
@@ -331,6 +389,25 @@ void CompositionalMultiphaseHybridFVM::assembleFluxTerms( real64 const dt,
       faceManager.getReference< array1d< globalIndex > >( faceDofKey );
     arrayView1d< integer const > const & faceGhostRank = faceManager.ghostRank();
 
+    // Get boundary face indicator (initialized during initializePostInitialConditionsPreSubGroups)
+    arrayView1d< integer const > const isBoundaryFaceView =
+      faceManager.getField< flow::isBoundaryFace >();
+    
+//    // Print isBoundaryFaceView for debugging
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp]: remove

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp.backup [hunk 4]
```diff
@@ -0,0 +1,380 @@
+/*
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp.backup]: remove?

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp (hunk 3 above)
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseHybridFVM.cpp]: I have cleaned the code of commented lines.

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels_impl.hpp [hunk 5]
```diff
@@ -1287,6 +1357,604 @@ FluxKernel::
...
+                                                                    facePhaseDens[0][0],
+                                                                    facePhaseMassDensLocal[0][0],
+                                                                    facePhaseVisc[0][0],
+                                                                    facePhaseEnthalpy[0][0],
+                                                                    facePhaseInternalEnergy[0][0],
+                                                                    facePhaseCompFracLocal[0][0],
+                                                                    faceTotalDens );
+
+        // Evaluate relative permeability at face saturation from flash calculation
+//      relPermWrapper.compute( facePhaseFrac[0][0], 0, 0 );
```
[the remark this answers is not in the document]
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels_impl.hpp:1786]: This follows the the same pattern as the FV compositional code.

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels_impl.hpp [hunk 6]
```diff
@@ -676,95 +682,107 @@ AssemblerKernelHelper::
...
+                                         dt,
+                                         divMassFluxes,
+                                         dDivMassFluxes_dElemVars,
+                                         dDivMassFluxes_dFaceVars,
+                                         dofColIndicesElemVars );
+
+
+      faceIndexMap[ifaceLoc] = numNonBoundaryFaces;
+      dofColIndicesFaceVars[numNonBoundaryFaces] = faceDofNumber[elemToFaces[ifaceLoc]];
+      numNonBoundaryFaces++;
```
[the remark this answers is not in the document]
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels_impl.hpp]: done.

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels_impl.hpp [hunk 7]
```diff
@@ -446,7 +449,7 @@ UpwindingHelper::
                                        localIndex ( & totalMobIds )[ NP ][ 3 ],
                                        localIndex ( & totalMobPos )[ NP ] )
 {
-  if( NP == 2 )
+  if constexpr ( NP == 2 )
```
[the remark this answers is not in the document]
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels_impl.hpp:452]: this should not be a problem

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels.hpp [hunk 8]
```diff
@@ -1138,12 +1234,19 @@ void kernelLaunchSelectorFaceSwitch( T value, LAMBDA && lambda )
   switch( value )
   {
     case 4:
-    { lambda( std::integral_constant< T, 4 >() ); return;}
+    {
+      return lambda( std::integral_constant< int, 4 >{} );
```
[the remark this answers is not in the document]
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/compositional/CompositionalMultiphaseHybridFVMKernels.hpp:1246]: it is already fixed.

File: src/coreComponents/physicsSolvers/fluidFlow/CMakeLists.txt [hunk 9]
```diff
@@ -152,36 +152,30 @@ set( fluidFlowSolvers_sources
...
   generateKernels( TEMPLATE ${kernelTemplateFile}
                    JSON kernelSpecs
                    KEY ${jsonKey}
                    HEADERS headerFiles
                    SOURCES sourceFiles )
-
-list(APPEND fluidFlowSolvers_headers ${headerFiles})
-list(APPEND fluidFlowSolvers_sources ${sourceFiles})
+  list(APPEND fluidFlowSolvers_headers ${headerFiles})
+  list(APPEND fluidFlowSolvers_sources ${sourceFiles})
```
[the remark this answers is not in the document]
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/CMakeLists.txt:174]: it is correct.

**Comment by corbett5** [comment]
About our recent conversation: if you want to stop the explicit template instantiation but keep the dispatch list I think you'll need to add a boolean argument to `generateKernels` to skip this loop

https://github.com/GEOS-DEV/GEOS/blob/2839ff525eba3f250357e7e1564ad5f9d22c6ade/src/cmake/GeosxMacros.cmake#L286

That's the only CMake change that I forsee

File: src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp [hunk 10]
```diff
@@ -0,0 +1,722 @@
...
+)xml";
+  return oss.str();
+}
+
+// Helper: copy arrayView1d<real64 const> to std::vector<real64>
+static inline std::vector< real64 > arrayViewToVector( arrayView1d< real64 const > & arr, localIndex n )
+{
+  arr.move( hostMemorySpace, false );
+  return std::vector< real64 >( arr.data(), arr.data() + n );
+}
```
Comment (castelletto1) [inline src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp:335]:
```suggestion
// Helper: copy arrayView1d<real64 const> to std::vector<real64>
static inline std::vector< real64 > arrayViewToVector( arrayView1d< real64 const > & arr )
{
  arr.move( hostMemorySpace, false );
  return std::vector< real64 >( arr.data(), arr.data() + arr.size() );
}
```
```

File: src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp [hunk 11]
```diff
@@ -0,0 +1,722 @@
...
+  out.reserve( static_cast< size_t >( nCells ) * static_cast< size_t >( nCols ) );
+  for( localIndex i = 0; i < nCells; ++i )
+  {
+    for( localIndex j = 0; j < nCols; ++j )
+    {
+      out.push_back( static_cast< real64 >( arr( i, j ) ) );
+    }
+  }
+  return out;
+}
```
Comment (castelletto1) [inline src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp]:
```suggestion
// Helper: copy a 2D view to flat vector (row-major by phase index), accept any 2D view type
template< typename View2D >
static inline std::vector< real64 > arrayView2dToVector( View2D & arr )
{
  arr.move( hostMemorySpace, false );
  localIndex nCells = arr.size( 0 );
  localIndex nCols = arr.size( 1 );
  std::vector< real64 > out;
  out.reserve( static_cast< size_t >( nCells ) * static_cast< size_t >( nCols ) );
  for( localIndex i = 0; i < nCells; ++i )
  {
    for( localIndex j = 0; j < nCols; ++j )
    {
      out.push_back( static_cast< real64 >( arr( i, j ) ) );
    }
  }
  return out;
}
```

File: src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp [hunk 12]
```diff
@@ -0,0 +1,722 @@
...
+    real64 x = centers[i][0];
+    real64 volume = volumes[i];
+    real64 pNumeric = p_h[i];
+    real64 pExact = 2.0 * (1.0 - x) + 1.0 * x;
+    l2Error_ReduceSum += (pNumeric - pExact) * (pNumeric - pExact) * volume;
+    totalVolume_ReduceSum += volume;
+  } );
+
+  real64 const data[2] = { l2Error_ReduceSum.get(), totalVolume_ReduceSum.get() };
+  real64 l2Error = std::sqrt( data[0] ) / data[1];
```
Comment (castelletto1) [inline src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp]: I would compute the relative error as 

$$
E_{\mathrm{rel}} = 
\frac{
  \sqrt{\displaystyle\sum_i (p_i^{\mathrm{num}} - p_i^{\mathrm{exact}})^2 * V_i}
}{
  \sqrt{\displaystyle\sum_i (p_i^{\mathrm{exact}})^2 * V_i}
}
$$


```suggestion
  RAJA::ReduceSum< parallelDeviceReduce, real64 > squaredPressureError_ReduceSum( 0.0 );
  RAJA::ReduceSum< parallelDeviceReduce, real64 > squaredExactPressure_ReduceSum( 0.0 );
  localIndex const n_cells  = subRegion.size();
  forAll< geos::parallelDevicePolicy<> >( n_cells, [=] GEOS_HOST_DEVICE ( localIndex const i )
  {
    real64 const x = centers[i][0];
    real64 const cellVolume = volumes[i];
    real64 const pNumeric = p_h[i];
    real64 const pExact = 2.0 * (1.0 - x) + 1.0 * x;
    squaredPressureError_ReduceSum += (pNumeric - pExact) * (pNumeric - pExact) * cellVolume;
    squaredExactPressure_ReduceSum += pExact * pExact * cellVolume;
  } );

// Gather global reductions
real64 const squaredPressureError = squaredPressureError_ReduceSum.get();
real64 const squaredExactPressure = squaredExactPressure_ReduceSum.get();

// Compute pressure relative L2 error
real64 const normalizedL2Error = std::sqrt(squaredPressureError) / std::sqrt(squaredExactPressure);
```

File: src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp [hunk 13]
```diff
@@ -0,0 +1,722 @@
...
+  std::string meshFile = GetParam();
+  if( meshFile == std::string( "polyhedral_voronoi_regular.vtu" ) )
+  {
+    EXPECT_NEAR( l2Error, 0.0, PRESSURE_L2_TOLERANCE );
+  }
+  else
+  {
+    EXPECT_GT( l2Error, PRESSURE_L2_TOLERANCE );
+  }
+}
```
Comment (castelletto1) [inline src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp]: See [suggestion below](https://github.com/GEOS-DEV/GEOS/pull/3765/files#r2482954535) for `CompositionalMFDNonTPFAExactnessTest`

**Review by castelletto1 — APPROVED** [review]
A couple of small suggestions, thanks!

File: src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp (hunk 13 above)
Comment (OmarDuran) [reply src/coreComponents/integrationTests/fluidFlowTests/testCompositionalMultiPhaseMFDPolyhedral.cpp]: Thank you, Nicola. Your suggestion was turned into a helper function and is now called when appropriate.

File: src/coreComponents/physicsSolvers/fluidFlow/kernelSpecs.json [hunk 14]
```diff
@@ -24,125 +24,6 @@
                    "generation call). Duplicate entries are unified internally" ]
   },
 
-  "SinglePhaseHybridFVMKernels": {
```
Comment (rrsettgast) [inline src/coreComponents/physicsSolvers/fluidFlow/kernelSpecs.json:27]: I don't think you want to delete this. perhaps comment it out for when you will come back and change the strategy to autogenerate the dispatch. I would gently recommend that you do that in this PR, but it is your call.

**Review by rrsettgast — APPROVED** [review]
Just make sure to create an issue that will result in the auto generation of the dispatch.

File: src/coreComponents/physicsSolvers/fluidFlow/kernelSpecs.json (hunk 14 above)
Comment (OmarDuran) [reply src/coreComponents/physicsSolvers/fluidFlow/kernelSpecs.json:27]: Based on our communication, I created the following issue: https://github.com/GEOS-DEV/GEOS/issues/3894
